sql/core/src/main/scala/org/apache/spark/sql/ArrowConverters.scala

```diff
@@ -83,53 +134,41 @@ private[sql] object ArrowConverters {
   }
 
   /**
-   * Transfer an array of InternalRow to an ArrowRecordBatch.
+   * Iterate over InternalRows and write to an ArrowRecordBatch.
    */
-  private[sql] def internalRowsToArrowRecordBatch(
-      rows: Array[InternalRow],
+  private def internalRowIterToArrowBatch(
+      rowIter: Iterator[InternalRow],
       schema: StructType,
-      allocator: RootAllocator): ArrowRecordBatch = {
-    val fieldAndBuf = schema.fields.zipWithIndex.map { case (field, ordinal) =>
-      internalRowToArrowBuf(rows, ordinal, field, allocator)
+      allocator: RootAllocator,
+      initialSize: Option[Int]): ArrowRecordBatch = {
+
+    val columnWriters = schema.fields.zipWithIndex.map { case (field, ordinal) =>
+      ColumnWriter(ordinal, allocator, field.dataType)
+        .init(initialSize)
+    }
+
+    rowIter.foreach { row =>
+      columnWriters.foreach { writer =>
+        writer.write(row)
+      }
+    }
+
+    val fieldAndBuf = columnWriters.map { writer =>
+      writer.finish()
     }.unzip
     val fieldNodes = fieldAndBuf._1.flatten
     val buffers = fieldAndBuf._2.flatten
 
-    val recordBatch = new ArrowRecordBatch(rows.length,
+    val rowLength = if(fieldNodes.nonEmpty) fieldNodes.head.getLength else 0
```

Comment:
It will probably never be empty, but might be good to keep the check just in case